ons-spark/spark-concepts/exercises.md

```diff
@@ -0,0 +1,233 @@
+    dplyr::summarise(animal_count = n())
+
+# Join this back to the rescue_with_pop DF
+rescue_with_pop <- rescue_with_pop %>%
+    sparklyr::left_join(animal_count, by="animal_group")
+
+# Filter out animals with small counts
+rescue_with_pop <- rescue_with_pop %>%
+    sparklyr::filter(animal_count >= 5)
+
+# Sort the final data
+rescue_with_pop <- rescue_with_pop %>%
+    dplyr::arrange(desc(total_cost))
+
+# Get the top 10 and row count
+top_10 <- rescue_with_pop %>% head(10) %>% sparklyr::collect()
+row_count <- rescue_with_pop %>% sparklyr::sdf_nrow()
+```
+````
+<details>
+<summary>Answer</summary>
+
+Using the ideas from the Optimising and Avoiding Shuffle chapter, we can:
+- Minimise actions: many of the `.show()`/`head(n) %>% collect` and `.count()`/`sdf_nrow()` actions are not needed. Just keep the ones at the end.
+- Caching: The two actions at the end are needed, so to avoid full recalculation of the DF, use a cache.
+- Broadcast join: Change the join from a sort merge join to a broadcast join. Here the broadcast is forced with the hint `F.broadcast()`/`sdf_broadcast()`, but you could also turn on automatic broadcasting by setting the `spark.sql.autoBroadcastJoinThreshold` in the config, or removing it altogether to use the default.
+- Avoid unnecessary sorting: Many of the sorts are not needed; remove all apart from the last one.
+- Window functions: Change the group by and join to use a window function.
+
+The code is now much shorter and neater, and will run much faster. Here we have chained operations together; this will not affect the efficiency of the code, but does make it easier to read.
+
+You could eliminate the need for Spark altogether and just use pandas or base R/dplyr, providing that the source data is small enough.
+
+````{tabs}
+```{code-tab} py
+import yaml
+from pyspark.sql import SparkSession, functions as F
+from pyspark.sql.window import Window
+
+with open("ons-spark/config.yaml") as f:
+    config = yaml.safe_load(f)
+    
+spark = (SparkSession.builder.master("local[2]")
+         .appName("exercises")
+         # Disable broadcast join by default
+         .config("spark.sql.autoBroadcastJoinThreshold", -1)
+         .getOrCreate())
+
+rescue_path = config["rescue_path"]
+population_path = config["population_path"]
+
+# Read source data
+rescue = (spark.read.parquet(rescue_path)
+          .select("incident_number", "animal_group", "total_cost", "postcode_district"))
+population = spark.read.parquet(population_path)
+
+rescue_with_pop = (
+    rescue
+    # Join population data to rescue data, using a broadcast join
+    .join(F.broadcast(population), on="postcode_district", how="left")
+
+    # Use a window function to get the count of each animal group
+    .withColumn("animal_count",
+                F.count("incident_number")
+                .over(Window.partitionBy("animal_group")))
+
+    # Filter out animals with small counts
+    .filter(F.col("animal_count") >= 5)
```

Comment:
Only thing I could think of adding here is to move the window and filter above the join. Perhaps Spark would do that automatically (would have to check the execution plan), and it wouldn't be a large gain in this example either. But the idea is to filter before the join so that there are fewer rows to join.

Reply:
I just checked when testing the code runs- Spark doesn't move the window and filter in front of the join automatically. I then copied the code and moved the join later in the copy, I ran both options a couple of times to see if there was an improvement in performance, not much in it on this scale really.

Reply:
It's hard to see performance improvements on a such a small DF, but the principle is there. I've added a small section on this to the Shuffling article and changed the order in the exercise result.